pywebdriver/plugins/mettler_toledo_8217.py

```diff
@@ -0,0 +1,109 @@
+
+class MettlerToledo8217ScaleDriver(AbstractScaleDriver):
+
+    def __init__(self, config, *args, **kwargs):
+        super().__init__(config, *args, **kwargs)
+        self.vendor_product = "mettler_toledo_8217"
+
+    @property
+    def _port(self):
+        return self.config.get(
+            "scale_driver", "port", fallback="/dev/ttyS0"
+        )
+
+    @property
+    def _baudrate(self):
+        return self.config.getint(
+            "scale_driver", "baudrate", fallback=9600
+        )
+
+    @property
+    def weight(self):
+        """Return the last reported weight of the scale."""
+        with self.data_lock:
+            return self.data.get("value", 0)
+
+    @property
+    def scale_status(self):
+        """Return the last reported status of the scale."""
+        with self.data_lock:
+            return self.data.get("status", "ERROR")
+
+    def acquire_data(self, connection):
+        """Acquire data over the connection."""
+        buffer = b""
+        stx = False
+        # ask for weight data
+        connection.write(b"W")
+        while True:
+            c = connection.read(1)
+            if not c:
+                # timeout
+                raise serial.SerialTimeoutException()
+            if c == b"\x02":
+                # start of answer
+                stx = True
+                buffer = b""
+            elif c == b"\r":
+                # end of answer
+                if not stx:
+                    continue
+                break
+            else:
+                buffer += c
+        match = ANSWER_RE.match(buffer)
+        if match is None:
+            raise ValueError("serial readout was not valid")
+        matchdict = match.groupdict()
+        _logger.debug(matchdict)
+        status = matchdict["status"]
+        weight = matchdict["weight"]
+        result = self.data.copy()
+        if weight is not None:
+            result.update({"value": float(weight), "status": "FIXED"})
+            return result
+        if not isinstance(status, bytes):
+            return result
+        status_byte = int.from_bytes(status, byteorder="big")
+        if status_byte & 0b1:
+            # in motion
+            result.update({"status": "ACQUIRING"})
+        elif status_byte & 0b110:
+            result.update({"status": "ERROR"})
+        return result
+
+    def establish_connection(self):
+        """Establish a connection. The connection must be a context manager."""
+        return serial.Serial(
+            port=self._port,
+            baudrate=self._baudrate,
+            parity=serial.PARITY_EVEN,
+            stopbits=serial.STOPBITS_ONE,
+            bytesize=serial.SEVENBITS,
+            timeout=1,
+        )
+
+    def is_connection_active(self, connection):
+        """Ascertain whether the connection is active and healthy."""
+        if not connection or not connection.isOpen():
+            return False
+        try:
+            connection.read(1)
```

Comment:
This is an awkwardness in the implementation. I wrote this first in `def run()` without writing any of the actual business logic of the invoked methods:

```python
while True:
    try:
        data = self.acquire_data(connection)
        with self.data_lock:
            self.data = data
        time.sleep(self.poll_interval)
    except Exception:
        _logger.exception("error during acquiring of data")
        if not self.is_connection_active(connection):
            # Force-close the connection.
            exit_stack.close()
            self.active = False
            break
        # While connection is still active, try again.
        continue
```

These were the assumptions:

- `acquire_data` sometimes just throws an error because some data was corrupted over the serial connection somewhere.
- These errors are not indicative of a problem with the connection.
- Re-starting the connection is slow/undesirable/whatever.

I got these assumptions from my reading of your protocol code implementation :) Maybe they are wrong, but reading code is hard.

In actual fact, I've not yet hit 'error during acquiring of data' yet without unplugging the scale.

Back to the awkwardness of the implementation: `serial.Serial` has no way of checking the health of the connection. `isOpen()` exists, but it can return `True` even when the plug isn't in the computer, so that's not very useful. So I just figured that attempting to read from the serial connection is the only way of knowing for sure.

I can remove `is_connection_active` altogether and do as you suggest. I don't know if that's useful in the abstract general case.

Reply:
We can't check for that exception, because it's implementation-specific. I think it's `serial.SerialException`, but I'm not sure.

We could do this, though. We just just need to raise a custom exception à la `ScaleConnectionError` in the implementation code when we get a `SerialException`, and check for that in the abstract code. But I'm not sure that this makes the code better. I think the current architecture is more explicit; it's just a little awkward in implementation.

Reply:
although it might make the architecture a little more complex, i think that your last proposition makes a lot of sense, and i like it: it’s always better to catch a specific exception than all of them.

Reply:
We'd still need to catch all other exceptions and handle those, unless we require that `acquire_data` only ever raises `ScaleConnectionError`, which is a difficult requirement in Python, or unless we're fine with the program aborting on an unexpected error in `acquire_data`.

I'll write something and we'll review it afterwards :)

Reply:
[6f0d4e4](https://github.com/akretion/pywebdriver/pull/102/commits/6f0d4e405e2f594f202118ca2ebef2abac5a1030)